README.md
=========

Software package providing segmentation nodes based on various techniques. Masked image or mask is provided, and can be tuned via dynamic_reconfigure. More information can be found in http://wiki.ros.org/pal_vision_segmentation. 


### Segmentation based on color histogram

Hereafter an example based on images recorded in a rosbag is presented.

#### Play the rosbag

Play the rosbag provided as example in pal_vision_segmentation:

```
rosbag play `rospack find pal_vision_segmentation`/etc/pringles.bag --loop
```

which publishes images of a pringles pot:

<img src="https://raw.github.com/pal-robotics/pal_vision_segmentation/hydro-devel/etc/pringles_example.png" />
    
#### Object template

In order to segment the pringles pot based on its color an image template like the following one is required:

<img src="https://raw.github.com/pal-robotics/pal_vision_segmentation/hydro-devel/etc/pringles_template.png" />
    
#### histogram_segmentation node

Launch the 'histogram_segmentation' node as follows so that the appropriate object template is used:

```
rosrun pal_vision_segmentation histogram_segmentation `rospack find pal_vision_segmentation`/etc/pringles_template.png image:=/stereo/  left/image _dilate_iterations:=5 _erode_iterations:=1    
```

#### Visualize the segmentation mask

```
rosrun image_view image_view image:=/histogram_segmentation/mask
```
    
<img src="https://raw.github.com/pal-robotics/pal_vision_segmentation/hydro-devel/etc/pringles_mask.png" />

#### Visualize the segmented image:

```
rosrun image_view image_view image:=/histogram_segmentation/image_masked
```
    
<img src="https://raw.github.com/pal-robotics/pal_vision_segmentation/hydro-devel/etc/pringles_segmented.png" />



CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package pal_vision_segmentation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Forthcoming
-----------
* Add depth_segmentation node
* Do not show images unless explicit argument in unit test
* Remove ros init from test
* Timestamp is taken from the original image
* Contributors: Bence Magyar, Jordi Pages, Mariia Dmitrieva

0.0.1 (2014-05-30)
------------------
* Cleanup and readme before release
* First catkin version of pal_vision_segmentation
* Contributors: Bence Magyar, Jordi Pages


test/histogram_segmentation_unit_test.cpp
#include <gtest/gtest.h>
#include <ros/ros.h>
#include <ros/package.h>

#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>


#include <pal_vision_segmentation/histogram.h>


void backProjectionTest(const std::string& imageFileName,
                        const std::string& templateFileName,
                        int dilateIterations, int dilateSize,
                        int erodeIterations, int erodeSize,
                        cv::Mat& mask)
{
  cv::Mat templateImg;

  std::cout << "Trying to read image from file: " << imageFileName << std::endl;
  std::cout << "Trying to read template from file: " << templateFileName << std::endl;

  templateImg = cv::imread(templateFileName, 1);

  cv::Mat image;
  image = cv::imread(imageFileName, 1);

  cv::MatND hist;
  pal_vision_util::calcHSVHist(templateImg, hist);
  cv::normalize(hist, hist, 0, 255, cv::NORM_MINMAX, -1, cv::Mat()); //set histogram bin values in [0, 255]
  cv::threshold(hist, hist, 128, 255, cv::THRESH_BINARY);            //all those bin values > threshold are set to 255, otherwise 0

  pal_vision_util::showHist(hist, false);

  int threshold = 254;
  pal_vision_util::backProject(image,
                               hist,
                               threshold,
                               mask,
                               dilateIterations, dilateSize, //dilate iterations and size
                               erodeIterations, erodeSize);

  cv::Mat imgMasked;
  image.copyTo(imgMasked, mask);

  cv::namedWindow("image");
  cv::imshow("image", image);

  cv::namedWindow("back-projection");
  cv::imshow("back-projection", mask);

  cv::namedWindow("image masked");
  cv::imshow("image masked", imgMasked);

  cv::waitKey(5000);

  bool ok = true;
  EXPECT_TRUE( ok );
}


TEST(histogram, test_back_projection)
{
  cv::Mat mask;

  //////////////////////////////////////////////////////////////////////////////////////
  // Synthetic example
  //////////////////////////////////////////////////////////////////////////////////////
  backProjectionTest(ros::package::getPath("pal_vision_segmentation") + "/test/etc/histogram_segmentation_sample.jpg",
                     ros::package::getPath("pal_vision_segmentation") + "/test/etc/histogram_segmentation_blue_template.jpg",
                     0, 3, 0, 3,
                     mask);

  //check some pixels of the resulting mask
  EXPECT_TRUE( mask.at<unsigned char>(45, 54)   == 0 );
  EXPECT_TRUE( mask.at<unsigned char>(52, 208)  == 0 );
  EXPECT_TRUE( mask.at<unsigned char>(165, 90)  == 0 );
  EXPECT_TRUE( mask.at<unsigned char>(158, 230) == 255 );

  //////////////////////////////////////////////////////////////////////////////////////
  // Real example
  //////////////////////////////////////////////////////////////////////////////////////
  backProjectionTest(ros::package::getPath("pal_vision_segmentation") + "/test/etc/pringles_frame.jpg",
                     ros::package::getPath("pal_vision_segmentation") + "/test/etc/pringles_template2.png",
                     9, 5, 0, 3,
                     mask);

  //check some pixels of the resulting mask
  EXPECT_TRUE( mask.at<unsigned char>(55, 55)   == 0 );
  EXPECT_TRUE( mask.at<unsigned char>(221, 246) == 255 );
  EXPECT_TRUE( mask.at<unsigned char>(460, 630) == 255 );
  EXPECT_TRUE( mask.at<unsigned char>(240, 450) == 0 );
  EXPECT_TRUE( mask.at<unsigned char>(365, 8)   == 255 );
  EXPECT_TRUE( mask.at<unsigned char>(208, 142) == 255 );
  EXPECT_TRUE( mask.at<unsigned char>(220, 140) == 255 );
  EXPECT_TRUE( mask.at<unsigned char>(20, 220) == 255 );

}

// Run all the tests that were declared with TEST()
int main(int argc, char **argv)
{
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}


src/template_segmentation.cpp

/*
 * Software License Agreement (Modified BSD License)
 *
 *  Copyright (c) 2012, PAL Robotics, S.L.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of PAL Robotics, S.L. nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 */

/** 
  * @file template_segmentation.cpp
  * @author Bence Magyar
  * @date May 2012
  * @brief Template matchined based segmentation node
  */

// PAL headers
#include <pal_vision_segmentation/TemplateSegmentConfig.h>
#include <pal_vision_segmentation/image_processing.h>

// ROS headers
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <dynamic_reconfigure/server.h>

// OpenCV headers
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

/***Variables used in callbacks***/
image_transport::Publisher mask_pub;
int window_width;
int window_height;
int coeffs_to_cancel;
cv::Mat image_rect;
image_transport::Publisher image_pub;
image_transport::Publisher debug_pub;
cv::Mat target_template;
cv::Mat raw_template;
int offset_x, offset_y;
/***end of callback section***/

void imageCb(const sensor_msgs::ImageConstPtr& msg)
{
    if(image_pub.getNumSubscribers() > 0 ||
       mask_pub.getNumSubscribers() > 0 ||
       debug_pub.getNumSubscribers() > 0)
    {
        cv_bridge::CvImagePtr cv_ptr;
        try
        {
            cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
        }
        catch (cv_bridge::Exception& e)
        {
            ROS_ERROR("cv_bridge exception: %s", e.what());
            return;
        }

        image_rect = cv_ptr->image;

        double ticksBefore = cv::getTickCount();

        cv::Mat matched;
        cv::Mat dct_d;
        pal_vision_util::dctNormalization(image_rect, dct_d, coeffs_to_cancel);
        cv::matchTemplate(dct_d, target_template, matched, CV_TM_CCORR_NORMED);

        double min_val, max_val;
        cv::Point max_loc, min_loc;
        cv::minMaxLoc(matched, &min_val, &max_val, &min_loc, &max_loc, cv::Mat());

        cv::Mat mask;
        mask = cv::Mat::zeros(image_rect.rows, image_rect.cols, CV_8UC1);

        const int start_col = (max_loc.x + offset_x - window_width/2  < 0) ?
                              0 : (max_loc.x + offset_x - window_width/2);
        const int start_row = (max_loc.y + offset_y - window_height/2 < 0) ?
                              0 : (max_loc.y + offset_y - window_height/2);
        const int end_col = (window_width/2 + max_loc.x + offset_x > mask.rows) ?
                            mask.rows : (window_width/2 + offset_x + max_loc.x);
        const int end_row = (window_height/2 + max_loc.y + offset_y > mask.cols) ?
                            mask.cols : (window_height/2 + max_loc.y + offset_y);
        for(int i=start_row; i<end_row; ++i)
            for(int j=start_col; j<end_col; ++j)
                mask.at<uchar>(i,j) = 255;

        if(mask_pub.getNumSubscribers() > 0)
        {
            cv_bridge::CvImage mask_msg;
            mask_msg.header = msg->header;
            mask_msg.encoding = sensor_msgs::image_encodings::MONO8;
            mask_msg.image = mask;
            mask_pub.publish(mask_msg.toImageMsg());
        }

        if(image_pub.getNumSubscribers() > 0)
        {
            cv::Mat masked;
            image_rect.copyTo(masked, mask);
            cv_bridge::CvImage masked_msg;
            masked_msg.header = msg->header;
            masked_msg.encoding = sensor_msgs::image_encodings::BGR8;
            masked_msg.image = masked;
            masked_msg.header.stamp = ros::Time::now();
            image_pub.publish(*masked_msg.toImageMsg());
        }

        //DEBUG
        if(debug_pub.getNumSubscribers() > 0)
        {
            cv::rectangle(dct_d, cv::Rect(start_col, start_row, window_width, window_height), cv::Scalar(255,0,0), 2);
            cv::circle(dct_d, cv::Point(max_loc.x + offset_x, max_loc.y + offset_y), 5, cv::Scalar(255,0,0), 1);
            cv_bridge::CvImage debug_msg;
            debug_msg.header = msg->header;
            debug_msg.encoding = sensor_msgs::image_encodings::TYPE_8UC3;
            debug_msg.image = dct_d;
            debug_pub.publish(*debug_msg.toImageMsg());
        }

        ROS_INFO("imageCb runtime: %f ms",
                 1000*(cv::getTickCount() - ticksBefore)/cv::getTickFrequency());
    }
}

void reconf_callback(pal_vision_segmentation::TemplateSegmentConfig &config, uint32_t level)
{
    window_width = config.window_width;
    window_height = config.window_height;
    coeffs_to_cancel = config.coeffs_to_cancel;

    pal_vision_util::dctNormalization(raw_template, target_template, coeffs_to_cancel);
}

int main(int argc, char *argv[] )
{
    ros::init(argc, argv, "template_segmentation");
    ros::NodeHandle nh("template_segmentation");
    nh.param<int>("window_width", window_width, 10);
    nh.param<int>("window_height", window_height, 10);
    nh.param<int>("coeffs_to_cancel", coeffs_to_cancel, 2);

    if(argc < 2)
    {
        ROS_ERROR("Template segmentation needs a template to match. Please provide it as a command line argument.");
        return -1;
    }

    std::string template_path(argv[1]);
    ROS_INFO("%s", template_path.c_str());
    raw_template = cv::imread(template_path);
    pal_vision_util::dctNormalization(raw_template, target_template, coeffs_to_cancel);

    offset_x = (target_template.rows-1)/2;
    offset_y = (target_template.cols-1)/2;

    image_transport::ImageTransport it(nh);
    image_transport::Subscriber rect_sub = it.subscribe("/image", 1, &imageCb);
    image_pub = it.advertise("image_masked", 1);
    mask_pub = it.advertise("mask", 1);
    debug_pub = it.advertise("debug",1);

    dynamic_reconfigure::Server<pal_vision_segmentation::TemplateSegmentConfig> server;
    dynamic_reconfigure::Server<pal_vision_segmentation::TemplateSegmentConfig>::CallbackType f;
    f = boost::bind(&reconf_callback, _1, _2);
    server.setCallback(f);

    ros::spin();
}



src/histogram_segmentation.cpp
/*
 * Software License Agreement (Modified BSD License)
 *
 *  Copyright (c) 2012, PAL Robotics, S.L.
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *   * Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *   * Redistributions in binary form must reproduce the above
 *     copyright notice, this list of conditions and the following
 *     disclaimer in the documentation and/or other materials provided
 *     with the distribution.
 *   * Neither the name of PAL Robotics, S.L. nor the names of its
 *     contributors may be used to endorse or promote products derived
 *     from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 */

/** 
  * @file histogram_segmentation.cpp
  * @author Bence Magyar
  * @date May 2012
  * @brief Histogram based segmentation node. \
  * Using histogram backprojection, normalization, thresholding and dilate operator. \
  * Erode operator can be added by uncommenting lines \
  * in histogram_segmentation.cpp and the corresponding cfg file.
  */

// PAL headers
#include <pal_vision_segmentation/HistogramSegmentConfig.h>
#include <pal_vision_segmentation/image_processing.h>
#include <pal_vision_segmentation/histogram.h>

// ROS headers
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <dynamic_reconfigure/server.h>

// OpenCV headers
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/video/video.hpp>

// Boost headers
#include <boost/filesystem.hpp>


/***Variables used in callbacks***/
image_transport::Publisher mask_pub;
int threshold;
int dilate_iterations;
int dilate_size;
int erode_iterations;
int erode_size;
int dark_pixels_threshold;
cv::Mat image_rect;
image_transport::Publisher image_pub;
image_transport::Publisher debug_pub;
cv::MatND target_hist;
/***end of callback section***/


/**
 * @brief removeDarkPixels given a BGR image and a binary image computed from it, removes the
 *        white pixel in the latter corresponding to low luminosity pixels in the former one.
 * @param[in] originalImg
 * @param[out] binaryImg
 * @param[in] threshold
 */
void removeDarkPixels(const cv::Mat& originalImg,
                      cv::Mat& binaryImg,
                      int threshold = 10)
{
  pal_vision_util::Image imgOriginalImg(originalImg), imgL, img_a, img_b;

  pal_vision_util::getLab(imgOriginalImg, imgL, img_a, img_b);

  //update header of cv::Mat L from imgL:
  cv::Mat L;
  L = cv::cvarrToMat(imgL.getIplImg());

  cv::Mat L_thresholded;
  L_thresholded = L.clone();
  cv::threshold(L, L_thresholded, threshold, 255, cv::THRESH_BINARY);
  cv::bitwise_and(binaryImg, L_thresholded, binaryImg);
}

void imageCb(const sensor_msgs::ImageConstPtr& msg)
{
    if(image_pub.getNumSubscribers() > 0 ||
       mask_pub.getNumSubscribers() > 0 ||
       debug_pub.getNumSubscribers() > 0)
    {
        cv_bridge::CvImagePtr cv_ptr;
        try
        {
            cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
        }
        catch (cv_bridge::Exception& e)
        {
            ROS_ERROR("cv_bridge exception: %s", e.what());
            return;
        }

        image_rect = cv_ptr->image;

        cv::Mat backProject;
        pal_vision_util::backProject(image_rect, target_hist, threshold, backProject);

        if ( dark_pixels_threshold > 0 )
          removeDarkPixels(image_rect, backProject, dark_pixels_threshold);

        cv::Mat mask, tmp1;

        if(dilate_iterations == 0 && erode_iterations == 0)
            mask = backProject;

        if(dilate_iterations > 0)
        {
            cv::dilate(backProject, erode_iterations == 0 ? mask: tmp1,
                       cv::Mat::ones(dilate_size, dilate_size, CV_8UC1),
                       cv::Point(-1, -1), dilate_iterations);
        }

        if(erode_iterations > 0)
        {
            cv::erode(dilate_iterations == 0 ? backProject : tmp1, mask,
                      cv::Mat::ones(erode_size, erode_size, CV_8UC1),
                      cv::Point(-1, -1), erode_iterations);
        }

        ros::Time now = msg->header.stamp; //ros::Time::now();

        if(mask_pub.getNumSubscribers() > 0)
        {
            cv_bridge::CvImage mask_msg;
            mask_msg.header = msg->header;
            mask_msg.header.stamp = now; //ros::Time::now();
            mask_msg.encoding = sensor_msgs::image_encodings::MONO8;
            mask_msg.image = mask;
            mask_pub.publish(mask_msg.toImageMsg());
        }

        if(image_pub.getNumSubscribers() > 0)
        {
            cv::Mat masked;
            image_rect.copyTo(masked, mask);
            cv_bridge::CvImage masked_msg;
            masked_msg.header = msg->header;
            masked_msg.encoding = sensor_msgs::image_encodings::BGR8;
            masked_msg.image = masked;
            masked_msg.header.stamp  = now; //ros::Time::now();
            image_pub.publish(*masked_msg.toImageMsg());
        }

        //DEBUG
        if(debug_pub.getNumSubscribers() > 0)
        {
            cv_bridge::CvImage debug_msg;
            debug_msg.header = msg->header;
            debug_msg.encoding = sensor_msgs::image_encodings::MONO8;
            debug_msg.image = backProject;
            debug_pub.publish(*debug_msg.toImageMsg());
        }
    }
}

void reconf_callback(pal_vision_segmentation::HistogramSegmentConfig &config, uint32_t level)
{
    threshold = config.threshold;
    dilate_iterations = config.dilate_iterations;
    dilate_size = config.dilate_size;
    erode_iterations = config.erode_iterations;
    erode_size = config.erode_size;
    dark_pixels_threshold = config.dark_pixels_threshold;
}

void computeHistogramFromFile(const std::string& template_path, cv::MatND& hist)
{
  if ( !boost::filesystem::is_directory(template_path) ) //if the given path is not a folder
  {
    if ( !boost::filesystem::is_regular_file(template_path) )
      throw std::runtime_error("The given path is not a file: " + template_path);

    cv::Mat raw_template = cv::imread(template_path);
    pal_vision_util::calcHSVHist(raw_template, hist);
    cv::normalize(hist, hist, 0, 255, cv::NORM_MINMAX, -1, cv::Mat());
    cv::threshold(hist, hist, 128, 255, cv::THRESH_BINARY);            //all those bin values > threshold are set to 255, otherwise 0
  }
  else //otherwise, it should be a folder
  {
    boost::filesystem::directory_iterator end_iter;
    std::vector<cv::Mat> images;
    for( boost::filesystem::directory_iterator dir_iter(template_path) ; dir_iter != end_iter ; ++dir_iter)
    {
      if ( boost::filesystem::is_regular_file(dir_iter->status()) )
      {
        std::string fullPathFileName = dir_iter->path().string();
        cv::Mat raw_template = cv::imread(fullPathFileName);
        if ( !raw_template.empty() )
          images.push_back(raw_template);
      }
    }
    if ( !images.empty() )
    {
      //pal_vision_util::calcHSVHist(images, hist);
      std::vector<cv::MatND> hists;
      //compute a separate histogram for each image
      pal_vision_util::calcHSVHist(images, hists);

      //create a new histogram where every bin takes the maximum value if any of the histogram's bin value is
      //larger than the threshold
      for (int h = 0; h < hists[0].rows; ++h)
        for (int s = 0; s < hists[0].cols; ++s)
        {
          int i = 0;
          bool found = false;
          while ( i < hists.size() && !found )
          {
           if ( hists[i].at<float>(h, s) > threshold )
           {
              hists[0].at<float>(h, s) = 255;
              found = true;
           }
           ++i;
          }
          if ( !found )
            hists[0].at<float>(h, s) = 0;
        }
      hist = hists[0].clone();
    }
    else
      throw std::runtime_error("No images found in the given path: " + template_path);
  }
}

int main(int argc, char *argv[] )
{
    ros::init(argc, argv, "histogram_segmentation"); //initialize with a default name
    ros::NodeHandle nh("~"); //use node name as sufix of the namespace
    nh.param<int>("threshold", threshold, 254);
    nh.param<int>("dilate_iterations", dilate_iterations, 9);
    nh.param<int>("dilate_size", dilate_size, 7);
    nh.param<int>("erode_iterations", erode_iterations, 0);
    nh.param<int>("erode_size", erode_size, 3);
    nh.param<int>("dark_pixels_threshold", dark_pixels_threshold, 15);

    if(argc < 2)
    {
        ROS_ERROR("Histogram segmentation needs a template to search for. Please provide it as a command line argument.");
        return -1;
    }

    std::string template_path(argv[1]);
    ROS_INFO_STREAM("Computing histogram from: " << template_path);

    computeHistogramFromFile( template_path, target_hist );

    image_transport::ImageTransport it(nh);
    image_transport::Subscriber rect_sub = it.subscribe("/image", 1, &imageCb);
    image_pub = it.advertise("image_masked", 1);
    mask_pub = it.advertise("mask", 1);
    debug_pub = it.advertise("debug",1);

    dynamic_reconfigure::Server<pal_vision_segmentation::HistogramSegmentConfig> server;
    dynamic_reconfigure::Server<pal_vision_segmentation::HistogramSegmentConfig>::CallbackType f;
    f = boost::bind(&reconf_callback, _1, _2);
    server.setCallback(f);

    ros::spin();
}
